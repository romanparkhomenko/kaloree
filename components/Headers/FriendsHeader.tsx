import React, { useState } from 'react';

// reactstrap components
import {
  Card,
  CardBody,
  CardTitle,
  Container,
  Row,
  Col,
  Form,
  FormGroup,
  Input,
  Table,
} from 'reactstrap';

import AddMeal from '../Forms/AddMeal';
import AddWeight from '../Forms/AddWeight';
import AddWorkout from '../Forms/AddWorkout';
import prisma from '../../lib/prisma';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const FriendsHeader: React.FC = ({ handleFriends }) => {
  const [search, setSearch] = useState('');
  const [searchResults, setSearchResults] = useState('');
  const [successMessage, setSuccessMessage] = useState('');
  const [errorMessage, setErrorMessage] = useState('');

  const searchForFriends = async (e: React.SyntheticEvent) => {
    e.preventDefault();
    setSuccessMessage('');
    setErrorMessage('');
    try {
      const body = { search };
      await fetch(`/api/friends/search`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      })
        .then(res => res.json())
        .then(response => {
          console.info(response);
          setSearchResults(response);
        })
        .catch(error => {
          setErrorMessage('Whoops, something went wrong');
          console.info(error);
        });
    } catch (error) {
      setErrorMessage('Whoops, something went wrong');
      console.error(error);
    }
  };

  const addFriend = async (e: React.SyntheticEvent, friendId) => {
    e.preventDefault();
    setSuccessMessage('');
    setErrorMessage('');
    try {
      const body = { friendId };
      await fetch(`/api/friends`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      })
        .then(res => res.json())
        .then(response => {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          const friend = searchResults.find(element => element.id === friendId);
          handleFriends(friend);
          console.info(response);
        })
        .catch(error => {
          setErrorMessage('Whoops, something went wrong');
          console.info(error);
        });
    } catch (error) {
      setErrorMessage('Whoops, something went wrong');
      console.error(error);
    }
  };

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return (
    <>
      <div className="header bg-gradient-dark pb-8 pt-5 pt-md-8">
        <Container fluid>
          <div className="header-body">
            {/* Card stats */}
            <Card>
              <CardBody>
                <Row className="align-items-end">
                  <Col lg="8" xl="8">
                    <label className="form-control-label" htmlFor="search">
                      Search For Friends
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="search"
                      placeholder="Search name or email"
                      onChange={e => setSearch(e.target.value)}
                      value={search}
                      type="text"
                    />
                  </Col>
                  <Col lg="4">
                    <button
                      className="btn btn-primary bg-gradient-primary mb-1 w-100"
                      onClick={e => searchForFriends(e)}
                    >
                      Search
                    </button>
                    {successMessage && (
                      <p className="text-center font-weight-700 text-info">{successMessage}</p>
                    )}
                    {errorMessage && (
                      <p className="text-center font-weight-700 text-warning">{errorMessage}</p>
                    )}
                  </Col>
                </Row>
                <Row className="mt-4 mb-1">
                  {searchResults && searchResults.length > 0 ? (
                    <Table className="align-items-center table-flush" responsive>
                      <thead className="thead-light">
                        <tr>
                          <th scope="col"></th>
                          <th scope="col">Name</th>
                          <th scope="col">Email</th>
                          <th scope="col">Add</th>
                        </tr>
                      </thead>
                      <tbody>
                        {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
                        {/*// @ts-ignore*/}
                        {searchResults.map(searchResult => (
                          <tr key={`user-${searchResult.id}`}>
                            <td>
                              <img
                                alt="..."
                                className="rounded-circle"
                                src={require('assets/img/theme/profile-placeholder.png')}
                                style={{ maxWidth: '25px' }}
                              />
                            </td>
                            <td>{searchResult.name ? searchResult.name : 'No Name'}</td>
                            <td>{searchResult.email}</td>
                            <td>
                              <button
                                className="btn btn-primary bg-gradient-orange w-auto"
                                onClick={e => addFriend(e, searchResult.id)}
                              >
                                Add friend
                              </button>
                            </td>
                          </tr>
                        ))}
                      </tbody>
                    </Table>
                  ) : (
                    <h3 className="text-center w-100 text-indigo my-1">No Results Found</h3>
                  )}
                </Row>
              </CardBody>
            </Card>
          </div>
        </Container>
      </div>
    </>
  );
};

export default FriendsHeader;
