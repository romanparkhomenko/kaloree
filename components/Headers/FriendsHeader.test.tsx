import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { render, fireEvent, waitFor, screen, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import FriendsHeader from './FriendsHeader';

describe('Friends Page Header', () => {
  const handleFriends = jest.fn();

  it('renders correctly', () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const tree = renderer.create(<FriendsHeader handleFriends={handleFriends} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('User can search friend', async () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    render(<FriendsHeader handleFriends={handleFriends} />);

    const searchInput = screen.getByPlaceholderText('Search name or email');
    const searchButton = screen.getByTestId('searchFriendsButton');

    fireEvent.change(searchInput, { target: { value: 'Roman' } });
    expect(searchInput).toHaveValue('Roman');

    const fakeSearchResponse = [
      {
        id: '1',
        name: 'Roman Parkhomenko',
        email: 'rsparkhomenko@gmail.com',
      },
    ];

    global.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        json: () => Promise.resolve(fakeSearchResponse),
      });
    });

    await act(async () => {
      fireEvent.click(searchButton);
    });

    // Confirm friend results show up
    await waitFor(() => {
      expect(screen.getByText('Add friend')).toBeVisible();
    });
  });

  it('User can add friend', async () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    render(<FriendsHeader handleFriends={handleFriends} />);

    const searchInput = screen.getByPlaceholderText('Search name or email');
    const searchButton = screen.getByTestId('searchFriendsButton');

    fireEvent.change(searchInput, { target: { value: 'Roman' } });
    expect(searchInput).toHaveValue('Roman');

    const fakeSearchResponse = [
      {
        id: '1',
        name: 'Roman Parkhomenko',
        email: 'rsparkhomenko@gmail.com',
      },
    ];

    global.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        json: () => Promise.resolve(fakeSearchResponse),
      });
    });

    await act(async () => {
      fireEvent.click(searchButton);
    });

    // Confirm friend results show up
    await waitFor(() => {
      expect(screen.getByText('Add friend')).toBeVisible();
    });

    const addFriendButton = screen.getByText('Add friend');
    await act(async () => {
      fireEvent.click(addFriendButton);
    });

    expect(handleFriends).toHaveBeenCalled();
  });
});
