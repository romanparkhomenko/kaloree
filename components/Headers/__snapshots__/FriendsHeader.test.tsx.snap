// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Friends Page Header renders correctly 1`] = `
<div
  className="header bg-gradient-dark pb-8 pt-5 pt-md-8"
>
  <div
    className="container-fluid"
  >
    <div
      className="header-body"
    >
      <div
        className="card"
      >
        <div
          className="card-body"
        >
          <div
            className="align-items-end row"
          >
            <div
              className="col-lg-8 col-xl-8"
            >
              <label
                className="form-control-label"
                htmlFor="search"
              >
                Search For Friends
              </label>
              <input
                className="form-control-alternative form-control"
                id="search"
                onChange={[Function]}
                placeholder="Search name or email"
                type="text"
                value=""
              />
            </div>
            <div
              className="col-lg-4"
            >
              <button
                className="btn btn-primary bg-gradient-primary mb-1 w-100"
                data-testid="searchFriendsButton"
                onClick={[Function]}
              >
                Search
              </button>
              
              
            </div>
          </div>
          <div
            className="mt-4 mb-1 row"
          >
            <h3
              className="text-center w-100 text-indigo my-1"
            >
              No Results Found
            </h3>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
